### ISSUE TITLE
Syntax highlighting Pass-2 - different behavior Nedit x Nedit-gn

### ISSUE THREAD

**Issue by jahodb** [eteran/nedit-ng#148]
I am using Nedit 5.6 now. I imported all settings including syntax highlighting by the help of command nedit-import to the Nedit-gn. Usually I use option "Pass-2 (parsing is deferred until text is exposed)" in the definition of patterns. There is different behaviour between Nedit 5.6 and Nedit-gn 2019.3/2019.4. Syntax highlighting is not updated after the text modification in Nedit-ng.
I have to switch highlighting off and on again to update it. Is it a bug or an intended feature? Thanks for the reply.

**Comment by eteran** [comment]
Hmm, sounds like a bug. The rules **should** behave exactly as they do in NEdit 5.6.

Can you provide your highlight patterns and an example of the syntax which is expected to be highlighted but isn't?

**Comment by jahodb** [comment]
I sent you the example files via eBox. As the contact e-mail I have used evan.teran@gmail.com. At this email the link for downloading will be sent automatically.Tar file contains:
- my configuration files:
  - config.ini
  - theme.xml
- example ASCII file example.inc - there are 6 lines - every odd one is correct and every even is not
- two screen shots with the result of the highlighting:
  - after the loading of the file: result_after_loading.png
  - after the editing: result_after_editing.png

**Comment by eteran** [comment]
Thanks, I'll check it out ASAP

**Comment by eteran** [comment]
So my understanding is that editing the two long lines should toggle the first portion of them to be green background or not based on if they are over 80 chars. Correct?

And that while the patterns seem to match correctly, they don't seem to be updated live?

**Comment by jahodb** [comment]
Exactly. If the length of the line is greater than 80 characters, the background for the first 80 characters is changed to green and the background of the rest of the line is red - see result_after_loading.png. If I edit the text, the syntax highlighting is updated for edited part of the string only - see result_after_editing.png. If I do the same in NEdit 5.6 the syntax highlighting is updated for the whole line. Thanks

**Comment by eteran** [comment]
Could you send me (it's easy to attach in here BTW) your classic nedit highlight rules just so it's easier for me to compare behaviors?

**Comment by eteran** [comment]
So a couple of things:

1. I manually made the same settings in classic nedit since it wasn't **that** much :-)

2. I am seeing what appears to be identical behavior in classic NEdit and NEdit-ng. Are you **sure** this worked in the original NEdit as you expect?

3. I agree that it is a bug... but it may be hard to figure out the fix since I was hoping to be able to compare the code paths of the new and old NEdits. If that ends up not being an option, it could prove to be challenging to figure out what's going on here. But I'll do my best. (I've been meaning to rework the highlighting engine entirely anyway, to be honest).

I'll attach a video showing what I mean in a little bit.

**Comment by eteran** [comment]
![simplescreenrecorder-2020-01-22_13 35 20](https://user-images.githubusercontent.com/571477/72923276-bac3e380-3d1c-11ea-97e1-83235f9effba.gif)

I really don't see any difference in behavior :-/

**Comment by jahodb** [comment]
My rc file for NEdit 5.6. is quite complex, therefore I wanted to prepare simpler example (there is added syntax highlighting for the *.inc file only). If I use this setup the NEdit 5.6. behaves like yours in the movie above = highlighting is not updated for the whole line. But if I use my original rc file NEdit behaves correctly in my opinion. Probably there is some other setting, that influences behaviour of the NEdit. I have not found yet, why there is the different behaviour for these two rc files. I attached both files. Thanks.

[nedit.rc.zip](https://github.com/eteran/nedit-ng/files/4101856/nedit.rc.zip)

**Comment by jahodb** [comment]
I think I have found it. There is small mistake in the pattern definition in the file nedit.rc_original - the Pattern type: "Pass 1 (applied ...)" is used for the last pattern named "ToFillIn". If I add this definition into file nedit.rc_simpler_examle the behaviour of the NEdit 5.6 is correct. It seems, usage of this single pattern causes update of the highlighting. The same process helps also in case of nedit-ng. If I do this change the highlighting of the file takes longer (example: file size 280MB - original time cca 2 s, new one 5 s).
[nedit.rc_simpler_examle_corrected.zip](https://github.com/eteran/nedit-ng/files/4103028/nedit.rc_simpler_examle_corrected.zip)

**Comment by eteran** [comment]
So I I understand correctly, then with that additional pattern, nedit-ng still acts like classic nedit?

**Comment by jahodb** [comment]
Exactly. The behaviour of both NEdits is the same now. Disadvantage of this "work around" is higher time consumption. I can live with this, but if it is possible to fix it, it would be better :-).

**Comment by eteran** [comment]
Just updating the category of this issue. It's not *technically* a bug since we are correctly replicating the behavior of classic nedit... but this should be fixed. So I'm considering it a "major enhancement"
